retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,593 @@
+				log.Errorf("Unable to watch services: %s", err)
+				return
+			}
+			kd.servicesResourceVersion = event.Service.ObjectMeta.ResourceVersion
+			events <- &event
+		}
+	}, retryDelay, kd.runDone)
+}
+
+func (kd *KubernetesDiscovery) updateService(service *kubernetes.Service, eventType kubernetes.EventType, ch chan<- *config.TargetGroup) {
+	kd.servicesMu.Lock()
+	defer kd.servicesMu.Unlock()
+	name := service.ObjectMeta.Name
+	namespace := service.ObjectMeta.Namespace
+	scrape := kd.checkScrape(service)
+	_, ok := kd.services[namespace][name]
+	switch eventType {
+	case kubernetes.Deleted:
+		if ok {
+			kd.deleteService(service, ch)
+		}
+	case kubernetes.Modified:
+		// Exists & no scrape - delete.
+		if ok && !scrape {
+			kd.deleteService(service, ch)
+			return
+		}
+
+		if scrape {
+			// Must be scrapable so let's add it to ensure labels are updated properly at least.
+			kd.addService(service, ch)
+		}
+	case kubernetes.Added:
+		if scrape {
+			kd.addService(service, ch)
+		}
+	}
+}
+
+func (kd *KubernetesDiscovery) deleteService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name}
+	ch <- tg
+	delete(kd.services[service.ObjectMeta.Namespace], service.ObjectMeta.Name)
+	if len(kd.services[service.ObjectMeta.Namespace]) == 0 {
+		delete(kd.services, service.ObjectMeta.Namespace)
+	}
+}
+
+func (kd *KubernetesDiscovery) addService(service *kubernetes.Service, ch chan<- *config.TargetGroup) {
+	namespace, ok := kd.services[service.ObjectMeta.Namespace]
+	if !ok {
+		namespace = make(map[string]*kubernetes.Service)
+		kd.services[service.ObjectMeta.Namespace] = namespace
+	}
+	namespace[service.ObjectMeta.Name] = service
+	endpointURL := fmt.Sprintf(kubernetesServiceEndpointsURL, service.ObjectMeta.Namespace, service.ObjectMeta.Name)
+	res, err := kd.client.Get(kd.conf.Server + endpointURL)
+	if err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+	if res.StatusCode != http.StatusOK {
+		log.Errorf("Failed to get service endpoints: %s", res.StatusCode)
+		return
+	}
+
+	var endpoints kubernetes.Endpoints
+	if err := json.NewDecoder(res.Body).Decode(&endpoints); err != nil {
+		log.Errorf("Error getting service endpoints: %s", err)
+		return
+	}
+
+	kd.updateServiceTargetGroup(service, &endpoints, ch)
+}
+
+func (kd *KubernetesDiscovery) updateServiceTargetGroup(service *kubernetes.Service, endpoints *kubernetes.Endpoints, ch chan<- *config.TargetGroup) {
+	tg := &config.TargetGroup{
+		Source: kubernetesSourceServicePrefix + ":" + service.ObjectMeta.Namespace + "/" + service.ObjectMeta.Name,
+		Labels: clientmodel.LabelSet{
+			kubernetesServiceNamespaceLabel: clientmodel.LabelValue(service.ObjectMeta.Namespace),
+			kubernetesServiceNameLabel:      clientmodel.LabelValue(service.ObjectMeta.Name),
+		},
+	}
+
+	for k, v := range service.ObjectMeta.Labels {
```

Comment:
> Service labels (which this uses) tend to be fairly constant, it is labels on pods, RCs, etc that tend to change more often from what I've seen.

I wasn't aware of the distinction. Will this give access to the pods labels too?

> The gauge isn't part of the Kubernetes label set, that comes from the scraped metric, right?

Yeah, people have done things like try and add label a for each service running on a machine to node exporter. I see this as the same, but at a different level.

> That sounds like a special case, not general case, to me. Also, they would necessarily be separate services with separate selectors & would want to get metrics on those separately, as well as aggregating over all related service types. Not sure why this wouldn't work correctly, other than the change every few minutes, which is a special case, not general case IMO.

The general case is being able to handle all of the special cases. We need to ensure that the default setup works (and labels changing every few minutes isn't working) for all reasonable use cases, while also allowing for specific use cases by additional configuration.

> Could add that as separate PR if it's agreed to go with this approach & update this SD to only use meta labels.

I can do that up as a separate PR for you.